### PR TITLE
Datastructure/board

### PR DESCRIPTION
**Pull request by yshrdbrn**
I added the basic methods that were needed. If you can think of anything else that we may need later, add it to the header.

### CODE REVIEW

File: datastructures/Board.cpp [hunk 1]
```diff
@@ -0,0 +1,160 @@
+//
+// Created by Yashar Dabiran on 2018-02-03.
```
Comment (jacobrs) [inline datastructures/Board.cpp]: Don't think we need to keep those, we can use git to track change ownership.

File: datastructures/Board.cpp [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+        board[2][i] = new Pawn(*pos, WHITE);
+
+        pos = new Position;
+        pos->column = (char)('A' + i - 1);
+        pos->row = 7;
+        board[7][i] = new Pawn(*pos, BLACK);
+    }
+
+    //Rook initialization
+    auto *pos = new Position;
```
Comment (jacobrs) [inline datastructures/Board.cpp]: I think you can bunch up the initialization lines:
```c++
auto *pos = new Position;
pos->row = 1;
pos->column = 'A';
```
Into a single line as such:
```c++
Position *pos = new Position({ 'A', 1 });
```

File: datastructures/Board.h [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    bool movePiece(Position &initPos, Position &destPos, Color player);
+
+    //Checks to see if the move is valid for the given player
+    bool isMoveValid(Position &initPos, Position &destPos, Color player);
+
+    std::vector<Position> validMovesOfPiece(Position &piecePos);
+
+    //Checks for the winner in the current state.
+    //Returns null if none of the players won, otherwise returns the color of the winner.
+    Color *whoWon();
```
Comment (jacobrs) [inline datastructures/Board.h:29]: Will this be the checkmate checker?

File: datastructures/pieces/Piece.h [hunk 4]
```diff
@@ -11,6 +11,7 @@ enum Color {
 struct Position {
     char column;
     int row;
+
```
Comment (jacobrs) [inline datastructures/pieces/Piece.h:14]: I saw in one of your commits that you added a == operator to the Position struct, is this where it should be?

**Review by jacobrs — COMMENTED** [review]
Overall looks good, couple comments. 👍

File: datastructures/Board.cpp (hunk 2 above)
Comment (yshrdbrn) [reply datastructures/Board.cpp]: I couldn't find the syntax for it. I'll change it. thanks

File: datastructures/Board.h (hunk 3 above)
Comment (yshrdbrn) [reply datastructures/Board.h:29]: Yeah

File: datastructures/Board.cpp (hunk 1 above)
Comment (yshrdbrn) [reply datastructures/Board.cpp]: Right. This was autogenerated. I forgot to delete it.

File: datastructures/pieces/Piece.h (hunk 4 above)
Comment (yshrdbrn) [reply datastructures/pieces/Piece.h:14]: Yeah, there was a problem so I deleted it temporarily. Now I added it again.

**Comment by jacobrs** [comment]
Changes look good, however I'm getting linking errors.
```
duplicate symbol __ZeqRK8PositionS1_ in:
    CMakeFiles/stocklion.dir/main.cpp.o
    CMakeFiles/stocklion.dir/datastructures/pieces/Piece.cpp.o
duplicate symbol __ZeqRK8PositionS1_ in:
    CMakeFiles/stocklion.dir/main.cpp.o
    CMakeFiles/stocklion.dir/datastructures/pieces/Pawn.cpp.o
duplicate symbol __ZeqRK8PositionS1_ in:
    CMakeFiles/stocklion.dir/main.cpp.o
    CMakeFiles/stocklion.dir/datastructures/pieces/Knight.cpp.o
duplicate symbol __ZeqRK8PositionS1_ in:
    CMakeFiles/stocklion.dir/main.cpp.o
    CMakeFiles/stocklion.dir/datastructures/pieces/Rook.cpp.o
duplicate symbol __ZeqRK8PositionS1_ in:
    CMakeFiles/stocklion.dir/main.cpp.o
    CMakeFiles/stocklion.dir/datastructures/pieces/Bishop.cpp.o
duplicate symbol __ZeqRK8PositionS1_ in:
    CMakeFiles/stocklion.dir/main.cpp.o
    CMakeFiles/stocklion.dir/datastructures/pieces/King.cpp.o
duplicate symbol __ZeqRK8PositionS1_ in:
    CMakeFiles/stocklion.dir/main.cpp.o
    CMakeFiles/stocklion.dir/datastructures/pieces/Queen.cpp.o
duplicate symbol __ZeqRK8PositionS1_ in:
    CMakeFiles/stocklion.dir/main.cpp.o
    CMakeFiles/stocklion.dir/datastructures/Board.cpp.o
ld: 8 duplicate symbols for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [stocklion] Error 1
make[1]: *** [CMakeFiles/stocklion.dir/all] Error 2
make: *** [all] Error 2
```
Is it just me?

**Comment by yshrdbrn** [comment]
Mine works fine, but I searched for the solution, it is because of the == operator defined in the header probably. Check to see if it works now.

**Comment by jacobrs** [comment]
Works! :+1:

**Review by jacobrs — APPROVED** [review]

File: datastructures/Board.cpp [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+        board[destPos.row][destPos.column - 'A' + 1] = board[initPos.row][initPos.column - 'A' + 1];
+        board[initPos.row][initPos.column - 'A' + 1] = nullptr;
+        return true;
+    }
+}
+
+bool Board::isMoveValid(Position &initPos, Position &destPos, Color player) {
+    auto piece = board[initPos.row][initPos.column - 'A' + 1];
+
+    if (piece->player != player)
```
Comment (benrs) [inline datastructures/Board.cpp:28]: What happens if we pass an invalid starting position (It starts as nullptr)

Comment (jacobrs) [reply datastructures/Board.cpp:28]: Agreed, perhaps if the position is invalid we can invalidate the move right away (nullptr check).

**Pull request merged**